File: 0014_LongestCommonPrefix/LongestCommonPrefix.cxx
#include<string>
#include<set>
#include<vector>
using namespace std;


class Solution1 {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if(strs.size() == 0) return "";
        if(strs.size() == 1) return strs[0];
        set<string> sorted_strs(strs.begin(), strs.end());
        auto lo = sorted_strs.begin();
        auto hi = --sorted_strs.end();
        
        string ans("");
        for( int i = 0; i< (*lo).size(); i++){
            if((*lo)[i] == (*hi)[i])
                ans.push_back((*lo)[i]);
            else break;
        }
        return ans;
    }
};
class Solution2 {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if (strs.size()==0)
            return "";
        auto lo = max_element(strs.begin(), strs.end());
        auto hi = min_element(strs.begin(), strs.end());
        auto pair = mismatch(lo->begin(), lo->end(), hi->begin());
        return string(lo->begin(), pair.first);
    }  
};


class Trie {
public:
    char next_ch = '-';
    Trie * next = nullptr;

    /** Initialize your data structure here. */
    Trie() {}
    void init(string str){
        Trie * root = this;
        for(auto && s : str){
            root->next_ch = s;
            root->next = new Trie();
            root = root->next;
        }
    }
    void CommonPrefix(string str){
        Trie * root = this;    
        for(auto && s : str){
            if(root->next_ch=='-') break;
            if(root->next_ch != s){
                root->next_ch = '-';
                break;
            }
            else{
                root = root->next;
            }      
        }
        root->next_ch = '-';
    }
};

class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if(strs.size() == 0 ) return "";
        if(strs.size() == 1) return strs[0];
        Trie * root = new Trie();
        root->init(strs[0]);
        for(auto && str : strs){
            if( str.size() == 0) return "";
            root->CommonPrefix(str);
        }
        string ans;
        while(root->next_ch != '-'){
            ans.push_back(root->next_ch);
            root = root->next;
        }
        return ans;
    }  
};

File: 0014_LongestCommonPrefix/LongestCommonPrefix.py
class Tries:
    def __init__(self):
        self.next_ch = '-'
        self.next_node = None
    
    
    def init(self, string:str):
        node = self
        for ch in string:
            node.next_ch=ch;
            node.next_node=Tries()
            node = node.next_node
    
    def CommonPrefix(self, string:str):
        node = self
        for ch in string:
            if node.next_ch == '-':
                break
            if node.next_ch != ch:
                node.next_ch='-'
                node.next_node = None
                break
            node = node.next_node
        node.next_ch='-'
        node.next_node=None
        

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs:
            return ''
        if len(strs) == 1: 
            return strs[0]
        root = Tries()
        root.init(strs[0])
        for string in strs[1:]:
            root.CommonPrefix(string)
        ans = ''
        while root.next_ch!='-':
            ans+=root.next_ch
            root=root.next_node
        return ans